File: mid/102. Binary Tree Level Order Traversal/102. Binary Tree Level Order Traversal.cpp
// https://leetcode.com/problems/binary-tree-level-order-traversal/solutions/2008729/o-n-time-beats-99-97-memory-speed-0ms-may-2022/
// Time complexity: O(n)
// Space complexity: O(h)

class Solution {
     public:
        vector<vector<int>> levelOrder(TreeNode* root) {
            if(root==NULL)
               return {};
            vector<vector<int>> ans; 
            queue<TreeNode*> q;
            q.push(root);
            TreeNode *temp;
            int len;
            while(!q.empty()){
               len=q.size();
		       vector<int> v;
               for(int i=0;i<len;i++){
                    temp=q.front();
                    q.pop();
                    v.push_back(temp->val);

                   if(temp->left) q.push(temp->left);
                   if(temp->right) q.push(temp->right);
                }
		        ans.push_back(v);
            }
          return ans;
        }
     };

File: mid/102. Binary Tree Level Order Traversal/102. Binary Tree Level Order Traversal.py
# https://leetcode.com/problems/binary-tree-level-order-traversal/solutions/5394346/best-approach-is-here-beats-100/
# Time complexity: (O(n)) where (n) is the number of nodes in the tree.
# Space complexity: (O(n)) The space required for the result list ans is proportional to the number of nodes. The queue can hold at most the number of nodes at the largest level, which in the worst case can be (O(n)).
# Beats 100%

class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
        
        ans = []
        queue = [root]
        
        while queue:
            level_size = len(queue)
            level = []
            
            for i in range(level_size):
                node = queue.pop(0)
                level.append(node.val)
                
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            
            ans.append(level)
        
        return ans